llms/mlx_lm/server.py

```diff
@@ -0,0 +1,277 @@
+                body["messages"],
+                tokenize=True,
+                add_generation_prompt=True,
+                return_tensors="np",
+            )
+        else:
+            prompt = convert_chat(body["messages"], body.get("role_mapping"))
+            prompt = _tokenizer.encode(prompt, return_tensors="np")
+        prompt = mx.array(prompt[0])
+        stop_words = body.get("stop", [])
+        stop_words = [stop_words] if isinstance(stop_words, str) else stop_words
+        stop_id_sequences = [
+            _tokenizer.encode(stop_word, return_tensors="np", add_special_tokens=False)[
+                0
+            ]
+            for stop_word in stop_words
+        ]
+        eos_token_id = _tokenizer.eos_token_id
+        max_tokens = body.get("max_tokens", 100)
+        stream = body.get("stream", False)
+        requested_model = body.get("model", "default_model")
+        temperature = body.get("temperature", 1.0)
+        top_p = body.get("top_p", 1.0)
+        if not stream:
+            tokens = list(
+                generate(
+                    prompt,
+                    _model,
+                    temperature,
+                    stop_id_sequences,
+                    eos_token_id,
+                    max_tokens,
+                    top_p=top_p,
+                )
+            )
+            text = _tokenizer.decode(tokens)
+            response = {
+                "id": chat_id,
+                "object": "chat.completion",
+                "created": int(time.time()),
+                "model": requested_model,
+                "system_fingerprint": f"fp_{uuid.uuid4()}",
+                "choices": [
+                    {
+                        "index": 0,
+                        "message": {
+                            "role": "assistant",
+                            "content": text,
+                        },
+                        "logprobs": None,
+                        "finish_reason": None,
+                    }
+                ],
+                "usage": {
+                    "prompt_tokens": len(prompt),
+                    "completion_tokens": len(tokens),
+                    "total_tokens": len(prompt) + len(tokens),
+                },
+            }
+
+            return response
+        else:
+            self.send_response(200)
+            self.send_header("Content-type", "text/event-stream")
+            self.send_header("Cache-Control", "no-cache")
+            self.end_headers()
+            accumulated_tokens = []
+            generated_text = ""
+            for token in generate(
+                prompt,
+                _model,
+                temperature,
+                stop_id_sequences,
+                eos_token_id,
+                max_tokens,
+                top_p=top_p,
+            ):
+                # This is a workaround because the llama tokenizer emits spaces during decoding token by token.
+                accumulated_tokens.append(token)
```

Comment:
This is a hacky workaround for the llama tokenizer. During decoding token by token, it emits spaces. However, when decoding the token list, it will add spaces properly.

Reply:
Related issue on transformers repo: https://github.com/huggingface/transformers/issues/22710